src/main/java/pl/edu/pw/mini/ingreedio/api/repository/impl/CustomizedProductRepositoryImpl.java

```diff
@@ -19,50 +25,152 @@ public class CustomizedProductRepositoryImpl implements CustomizedProductReposit
+
+        Criteria ratingFilteringCriteria = productCriteria.getMinRating() != null
+            ? Criteria.where("rating").gte(productCriteria.getMinRating())
+            : new Criteria();
+        Criteria phraseFilteringCriteria = new Criteria();
+
+        if (phraseKeywordsRegExp != null) {
+            phraseFilteringCriteria = new Criteria().orOperator(
+                Criteria.where("name").regex(phraseKeywordsRegExp),
+                Criteria.where("brand").regex(phraseKeywordsRegExp),
+                Criteria.where("shortDescription").regex(phraseKeywordsRegExp)
+            );
         }
-        if (criteria.getVolumeFrom() != null && criteria.getVolumeTo() != null) {
-            query.addCriteria(
-                Criteria.where("volume")
-                    .gte(criteria.getVolumeFrom())
-                    .lte(criteria.getVolumeTo()));
-        } else if (criteria.getVolumeFrom() != null) {
-            query.addCriteria(Criteria.where("volume").gte(criteria.getVolumeFrom()));
-        } else if (criteria.getVolumeTo() != null) {
-            query.addCriteria(Criteria.where("volume").lte(criteria.getVolumeTo()));
+
+        // Final operations List
+        List<AggregationOperation> finalQueryOperations = new ArrayList<>();
+
+        AggregationOperation filteringOperation = Aggregation.match(new Criteria().andOperator(
+            ingredientsFilteringCriteria,
+            brandsFilteringCriteria,
+            providerFilteringCriteria,
+            categoriesFilteringCriteria,
+            ratingFilteringCriteria,
+            phraseFilteringCriteria
+        ));
+
+        finalQueryOperations.add(filteringOperation);
+
+        // Stage 2: Prepare match score for each product (if there is match score sort operation)
+        if (productCriteria.getHasMatchScoreSortCriteria() != null
+            && productCriteria.getHasMatchScoreSortCriteria()
+            && phraseKeywordsRegExp != null) {
+
+            String addMatchScoreQuery =
+                "{\n"
+                + "  \"$addFields\": {\n"
+                + "    \"matchScore\": {\n"
+                + "      \"$add\": [\n"
+                + "        { \"$multiply\": [{ \"$size\": { \"$regexFindAll\": "
+                + "{ \"input\": \"$shortDescription\", \"regex\": /"
+                    + phraseKeywordsRegExp.pattern()
+                + "/, \"options\": 'i' } } }, 5] },\n"
+                + "        { \"$multiply\": [{ \"$size\": { \"$regexFindAll\": "
+                + "{ \"input\": \"$brand\", \"regex\": /" + phraseKeywordsRegExp.pattern()
+                + "/, \"options\": 'i' } } }, 10] },\n"
+                + "        { \"$multiply\": [{ \"$size\": { \"$regexFindAll\": "
+                + "{ \"input\": \"$name\", \"regex\": /" + phraseKeywordsRegExp.pattern()
+                + "/, \"options\": 'i' } } }, 15] }\n"
+                + "      ]\n"
+                + "    }\n"
+                + "  }\n"
+                + "}\n";
```

Comment:
I have a bone to pick with this fragment 🤓 I would do it like this:
1. Create query string as const with format placeholders and text block notation
```java 
private static final String MATCH_SCORE_QUERY = """
query here %1$s reusing %1$s the same formatting argument %1$s
"""
```
2. Replace this fragment with 
```java
String.format(MATCH_SCORE_QUERY, phraseKeywordsRegExp.pattern())
```

Reply:
It would be even greater if this query was in a resource file, but its not a 'must have'.

Reply:
I hoped that someone would give an idea how to do it better🙏🙏 thanks